### PR TITLE
build: Add multi-arch support

### PR DESCRIPTION
**Pull request by haoyu01**
This PR is to Add multi-arch support .
SHA1 MB patches for aarch64 are verified for the multi-arch patches

The PR is verified base on aarch64 platform .

### CODE REVIEW

File: include/memcpy_inline.h [hunk 1]
```diff
@@ -37,13 +37,18 @@
 #ifndef _MEMCPY_H_
 #define _MEMCPY_H_
 
+#if defined(__i386__) || defined(__x86_64__)
 #include "intrinreg.h"
+#endif
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply include/memcpy_inline.h]: I will change it .

[the remark this answers is not in the document]
Comment (haoyu01) [reply include/memcpy_inline.h]: I will change it .

File: Makefile.am [hunk 2]
```diff
@@ -19,20 +19,25 @@ perf_tests_extra=
...
+lsrc_aarch64=
+lsrc_base_aliases=
 unit_tests32=
 perf_tests32=
 
 # Include units
-include md5_mb/Makefile.am
 include sha1_mb/Makefile.am
+include mh_sha1/Makefile.am
+if !CPU_AARCH64
```
[the remark this answers is not in the document]
Comment (haoyu01) [reply Makefile.am]: @gbtucker , just make sure the commit can pass CI test . 
And yes , it is going to be paired with other commits .

File: make.inc [hunk 3]
```diff
@@ -44,7 +44,11 @@ host_cpu ?= $(shell uname -m | sed -e 's/amd/x86_/')
 arch ?= $(shell uname | grep -v -e Linux -e BSD )
 
 CC  = gcc
+ifeq ($(host_cpu),aarch64)
```
[the remark this answers is not in the document]
Comment (haoyu01) [reply make.inc]: OKay , I will change it

File: make.inc [hunk 4]
```diff
@@ -54,7 +58,11 @@ DEBUG_nasm = -g
 
 # Default arch= build options
 CFLAGS_    = -Wall
+ifeq ($(host_cpu),aarch64)
```
[the remark this answers is not in the document]
Comment (haoyu01) [reply make.inc]: Okay , I will change it

File: make.inc [hunk 5]
```diff
@@ -151,10 +159,25 @@ $(addsuffix .run,$(all_tests)): %.run : %
 # Other build rules
 msg = $(if $(DEBUG),DEBUG) $(patsubst 32,32-bit,$(host_cpu)) $D
 
+ifeq ($(host_cpu),aarch64)
```
[the remark this answers is not in the document]
Comment (haoyu01) [reply make.inc]: It was removed . 
And I add vpath setting in Makefile.unx

File: make.inc (hunk 3 above)
[the remark this answers is not in the document]
Comment (haoyu01) [reply make.inc]: I add below override 
`
ifeq ($(host_cpu)_$(arch),aarch64_)
arch = aarch64
endif
`

**Comment by haoyu01** [comment]
@gbtucker , just update the patch , please review it .

**Comment by haoyu01** [comment]
@gbtucker , just update the patch and please review it

**Comment by haoyu01** [comment]
Refresh Jira Link

**Comment by gbtucker** [comment]
It looks like a good port of the build scripts to be more arch generic.  Let's wait to pair this with some optimizations so we can test all of it.  Perhaps when you do we can rename to "Add multi-arch support" as this is what most of this patch is doing.

**Comment by haoyu01** [comment]
@gbtucker upload sha1_mb patches . Please review them .

File: sha1_mb/aarch64/sha1_mb_mgr_ce.c [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+#define LANE_IS_FINISHED(state,i)  	\
+	(((state->lens[i]&(~0xf))==0) && state->ldata[i].job_in_lane!=NULL)
+#define	LANE_IS_FREE(state,i)		\
+	(((state->lens[i]&(~0xf))==0) && state->ldata[i].job_in_lane==NULL)
+#define LANE_IS_INVALID(state,i)	\
+	(((state->lens[i]&(~0xf))!=0) && state->ldata[i].job_in_lane==NULL)
+void sha1_mb_mgr_init_ce(SHA1_MB_JOB_MGR * state)
+{
+	unsigned int i;
+	//~ state->unused_lanes = 0xf3210;
```
Comment (gbtucker) [inline sha1_mb/aarch64/sha1_mb_mgr_ce.c]: Should remove dead code.

File: sha1_mb/aarch64/sha1_mb_mgr_ce.c [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+		return -1;
+	}
+#if SHA1_MB_CE_MAX_LANES == 4
+	if (state->num_lanes_inuse == 4) {
+		len = min(min(state->lens[0], state->lens[1]),
+			  min(state->lens[2], state->lens[3]));
+		lane_idx = len & 0xf;
+		len &= ~0xf;
+
+		sha1_mb_ce_x4(state->ldata[0].job_in_lane,
```
Comment (gbtucker) [inline sha1_mb/aarch64/sha1_mb_mgr_ce.c]: Should add these functions only when they exist.

File: Makefile.am [hunk 8]
```diff
@@ -109,6 +109,7 @@ if USE_NASM
 endif
 if CPU_AARCH64
   as_filter = gcc -D__ASSEMBLY__
+  my_CFLAGS += -march=armv8-a+crypto
```
[the remark this answers is not in the document]
Comment (haoyu01) [reply Makefile.am]: done

File: include/sha1_mb.h [hunk 9]
```diff
@@ -443,6 +443,10 @@ void      sha1_mb_mgr_init_avx512_ni    (SHA1_MB_JOB_MGR *state);
 SHA1_JOB* sha1_mb_mgr_submit_avx512_ni  (SHA1_MB_JOB_MGR *state, SHA1_JOB* job);
 SHA1_JOB* sha1_mb_mgr_flush_avx512_ni   (SHA1_MB_JOB_MGR *state);
 
+void      sha1_mb_mgr_init_ce   (SHA1_MB_JOB_MGR *state);
```
[the remark this answers is not in the document]
Comment (haoyu01) [reply include/sha1_mb.h]: move it to sha1_mb_ctx_ce.c

File: sha1_mb/aarch64/sha1_mb_mgr_ce.c (hunk 6 above)
Comment (haoyu01) [reply sha1_mb/aarch64/sha1_mb_mgr_ce.c]: removed :)

File: sha1_mb/aarch64/sha1_mb_mgr_ce.c (hunk 7 above)
Comment (haoyu01) [reply sha1_mb/aarch64/sha1_mb_mgr_ce.c]: done . The functions were removed

**Comment by haoyu01** [comment]
@gbtucker , just update the code , please review .

**Comment by haoyu01** [comment]
@gbtucker I have changed the summary line . Please review it

**Pull request merged**